HLTrigger/HLTfilters/plugins/HLTDoubletDZ.cc

```diff
@@ -310,348 +342,271 @@ HLTDoubletDZ<T1, T2>::getCollections(edm::Event& iEvent,
     }
 
     return true;
-  } else
-    return false;
+  }
+
+  return false;
 }
 
 template <typename T1, typename T2>
-bool HLTDoubletDZ<T1, T2>::computeDZ(edm::Event& iEvent, T1Ref& r1, T2Ref& r2) const {
-  const reco::Candidate& candidate1(*r1);
-  const reco::Candidate& candidate2(*r2);
-  if (reco::deltaR(candidate1, candidate2) < minDR_)
-    return false;
-  if (std::abs(candidate1.vz() - candidate2.vz()) > maxDZ_)
-    return false;
-
-  return true;
+bool HLTDoubletDZ<T1, T2>::haveSameSuperCluster(T1 const& c1, T2 const& c2) const {
+  return (c1.superCluster().isNonnull() and c2.superCluster().isNonnull() and (c1.superCluster() == c2.superCluster()));
 }
 
 template <>
-bool HLTDoubletDZ<reco::RecoEcalCandidate, reco::RecoChargedCandidate>::computeDZ(edm::Event& iEvent,
-                                                                                  T1Ref& r1,
-                                                                                  T2Ref& r2) const {
-  edm::Handle<reco::ElectronCollection> electronHandle_;
-  iEvent.getByToken(electronToken_, electronHandle_);
-  if (!electronHandle_.isValid())
-    edm::LogError("HLTDoubletDZ") << "HLTDoubletDZ: Electron Handle not valid.";
-
-  if (reco::deltaR(*r1, *r2) < minDR_)
-    return false;
-  reco::Electron e1;
-  for (auto const& eleIt : *electronHandle_) {
-    if (eleIt.superCluster() == r1->superCluster())
-      e1 = eleIt;
-  }
-
-  const reco::RecoChargedCandidate& candidate2(*r2);
-  bool skipDZ = false;
-  if (minPixHitsForDZ_ > 0 && (e1.gsfTrack()->hitPattern().numberOfValidPixelHits() < minPixHitsForDZ_ ||
-                               candidate2.track()->hitPattern().numberOfValidPixelHits() < minPixHitsForDZ_))
-    skipDZ = true;
-  if (!skipDZ && std::abs(e1.vz() - candidate2.vz()) > maxDZ_)
-    return false;
-
-  return true;
```

Comment:
The filter logic confuses me a bit. If there is a requirement on `minPixHitsForDZ_`, but either one of the candidates fails it, then the DZ cut is skipped, and the pair of candidates is accepted. I think this is the intended behaviour, but better to ask for confirmation. @swagata87 , do you know more ?

Reply:
yes you are right. this is intended, as far as I am aware. 
It can happen that a perfectly good signal electron just don't happen to have hits in the first 1 or 2 layers of inner tracker. Reasons could be detector issue, inefficiency of the reconstruction algorithms we use, hard brem and so on..  
In that case, that electron might pass all other cuts in HLT but will likely fail this DZ cut. 
So DZ cut can introduce some unintended inefficiency at HLT. For dielectron paths with pT greater than 10-20ish GeV, the efficiency loss due to DZ is "tolerable". But for low pT electrons, the efficiency loss due to DZ can be very significant. So, `minPixHitsForDZ_ ` gives an opportunity to loosen the DZ requirement. It allows us to apply the DZ cut only when there are at least X hits in the inner tracker for both electrons; X being 3, or 4 depending on needs.